src/typechecker/FStar.TypeChecker.Core.fst

```diff
@@ -1002,10 +1005,21 @@ let rec check_relation (g:env) (rel:relation) (t0 t1:typ)
         if not (head_matches && List.length args0 = List.length args1)
         then maybe_unfold_and_retry t0 t1
         else (
-          handle_with
-            (check_relation g EQUALITY head0 head1 ;!
-             check_relation_args g EQUALITY args0 args1)
-            (fun _ -> maybe_unfold_side_and_retry Both t0 t1)
+          (* If we're proving equality and SMT queries are ok, 
+             defer things like `v.v1 == u.v1` to SMT
+             rather than matching on the v1 projector and
+             trying to prove `v == u` *)
+          if guard_ok &&
+            (rel=EQUALITY) && 
+            equatable g t0 &&
```

Comment:
Is the correct check `equatable g t0 && equatable g t1` or `equatable g t0 || equatable g t1` ... we do the latter at another place in this function.

Reply:
I've revised this